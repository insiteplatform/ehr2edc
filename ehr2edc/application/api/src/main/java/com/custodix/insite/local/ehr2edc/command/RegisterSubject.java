package com.custodix.insite.local.ehr2edc.command;

import static com.custodix.insite.local.ehr2edc.shared.annotations.security.Allow.Rule.ASSIGNED_INVESTIGATOR;

import java.time.LocalDate;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;

import com.custodix.insite.local.ehr2edc.infrastructure.validation.DateRange;
import com.custodix.insite.local.ehr2edc.shared.annotations.AuthorizationCorrelator;
import com.custodix.insite.local.ehr2edc.shared.annotations.SynchronizationCorrelator;
import com.custodix.insite.local.ehr2edc.shared.annotations.security.Allow;
import com.custodix.insite.local.ehr2edc.vocabulary.EDCSubjectReference;
import com.custodix.insite.local.ehr2edc.vocabulary.PatientCDWReference;
import com.custodix.insite.local.ehr2edc.vocabulary.StudyId;
import com.custodix.insite.local.ehr2edc.vocabulary.SubjectId;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;

public interface RegisterSubject {

	@Allow(ASSIGNED_INVESTIGATOR)
	Response register(@Valid @NotNull Request request);

	@JsonDeserialize(builder = Request.Builder.class)
	final class Request {
		@NotNull
		@Valid
		private final PatientCDWReference patientId;
		@SynchronizationCorrelator
		@AuthorizationCorrelator
		@NotNull
		@Valid
		private final StudyId studyId;
		@NotNull
		@Valid
		private final EDCSubjectReference edcSubjectReference;
		@NotNull
		@DateRange(min = "1950-01-01")
		private final LocalDate dateOfConsent;

		@NotNull
		private final String firstName;

		@NotNull
		private final String lastName;

		@NotNull
		@DateRange
		private final LocalDate birthDate;

		private Request(Builder builder) {
			patientId = builder.patientId;
			studyId = builder.studyId;
			edcSubjectReference = builder.edcSubjectReference;
			dateOfConsent = builder.dateOfConsent;
			firstName = builder.firstName;
			lastName = builder.lastName;
			birthDate = builder.birthDate;
		}

		public static Builder newBuilder() {
			return new Builder();
		}

		public static Builder newBuilder(Request copy) {
			Builder builder = new Builder();
			builder.patientId = copy.patientId;
			builder.studyId = copy.studyId;
			builder.edcSubjectReference = copy.edcSubjectReference;
			builder.dateOfConsent = copy.dateOfConsent;
			builder.lastName = copy.lastName;
			builder.firstName = copy.firstName;
			builder.birthDate = copy.birthDate;
			return builder;
		}

		public PatientCDWReference getPatientId() {
			return patientId;
		}

		public StudyId getStudyId() {
			return studyId;
		}

		public EDCSubjectReference getEdcSubjectReference() {
			return edcSubjectReference;
		}

		public LocalDate getDateOfConsent() {
			return dateOfConsent;
		}

		public String getFirstName() {
			return firstName;
		}

		public String getLastName() {
			return lastName;
		}

		public LocalDate getBirthDate() {
			return birthDate;
		}

		@JsonPOJOBuilder
		public static final class Builder {
			private PatientCDWReference patientId;
			private StudyId studyId;
			private LocalDate dateOfConsent;
			private String firstName;
			private String lastName;
			private LocalDate birthDate;
			private EDCSubjectReference edcSubjectReference;

			private Builder() {
			}

			public Builder withPatientId(PatientCDWReference patientCDWReference) {
				this.patientId = patientCDWReference;
				return this;
			}

			public Builder withStudyId(StudyId studyId) {
				this.studyId = studyId;
				return this;
			}

			public Builder withEdcSubjectReference(EDCSubjectReference edcSubjectReference) {
				this.edcSubjectReference = edcSubjectReference;
				return this;
			}

			public Builder withDateOfConsent(LocalDate dateOfConsent) {
				this.dateOfConsent = dateOfConsent;
				return this;
			}

			public Builder withFirstName(String val) {
				firstName = val;
				return this;
			}

			public Builder withLastName(String val) {
				lastName = val;
				return this;
			}

			public Builder withBirthDate(LocalDate val) {
				birthDate = val;
				return this;
			}

			public Request build() {
				return new Request(this);
			}
		}
	}

	final class Response {
		private final SubjectId subjectId;

		private Response(Builder builder) {
			subjectId = builder.subjectId;
		}

		public SubjectId getSubjectId() {
			return subjectId;
		}

		public static Builder newBuilder() {
			return new Builder();
		}

		public static final class Builder {
			private SubjectId subjectId;

			private Builder() {
			}

			public Builder withSubjectId(SubjectId subjectId) {
				this.subjectId = subjectId;
				return this;
			}

			public Response build() {
				return new Response(this);
			}
		}
	}
}