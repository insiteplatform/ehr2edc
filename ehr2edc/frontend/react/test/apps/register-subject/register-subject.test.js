import React from 'react';
import renderer from 'react-test-renderer';
import RegisterSubject from "../../../src/apps/register-subject/register-subject";
import sinon from 'sinon';
import readDataFromFile from "../../__helper__/sample-data-file";

const DATE_OF_CONSENT = "2019-04-12";

describe("Rendering the component", () => {
    'use strict';

    const spyOnFetchStudies = sinon.spy(RegisterSubject.prototype, 'fetchStudies');
    const spyOnRegister = sinon.spy(RegisterSubject.prototype, 'registerPatientForStudy');
    const spyOnMount = sinon.spy(RegisterSubject.prototype, 'componentDidMount');
    const spyOnUpdate = sinon.spy(RegisterSubject.prototype, 'componentDidUpdate');

    const patientStudyApi = {};

    const patientRegistrationApi =
        {
            studyId: {},
            registerSubjectRequest: {},
            put: function (_studyId, _body) {
                this.studyId = _studyId;
                this.registerSubjectRequest = JSON.parse(JSON.stringify(_body));
                return Promise.resolve({ status: 201 })
            },
        };


    const onClose = jest.fn();
    const DEFAULT_PATIENT_ID_SRC = "MASTER_PATIENT_INDEX";

    beforeEach(() => {
        spyOnFetchStudies.resetHistory();
        spyOnMount.resetHistory();
        spyOnUpdate.resetHistory();
        spyOnRegister.resetHistory();
    });

    test("The page is rendered", async () => {
        letWebserviceReturnStudyDataForPatient("allStudiesForPatient-noStudies-response");
        const app = renderer.create(
            <RegisterSubject
                patientStudyApi={patientStudyApi}
                patientRegistrationApi={patientRegistrationApi}
                patientSource={DEFAULT_PATIENT_ID_SRC}
                onClose={onClose}
            />
        );
        expect(spyOnMount.callCount).toEqual(1);
        expect(spyOnUpdate.callCount).toEqual(0);
        expect(spyOnFetchStudies.callCount).toEqual(1);

        await Promise.resolve();

        const tree = app.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("The page is rendered with a patientId", async () => {
        letWebserviceReturnStudyDataForPatient("allStudiesForPatient-response");
        const patientId = "PID_001";
        const app = renderer.create(
            <RegisterSubject
                patientStudyApi={patientStudyApi}
                patientRegistrationApi={patientRegistrationApi}
                patientSource={DEFAULT_PATIENT_ID_SRC}
                patientId={patientId}
                onClose={onClose}
            />
        );
        expect(spyOnMount.callCount).toEqual(1);
        expect(spyOnUpdate.callCount).toEqual(0);
        expect(spyOnFetchStudies.callCount).toEqual(1);

        await Promise.resolve();

        const tree = app.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("The page is rendered changing the patientId", async () => {
        letWebserviceReturnStudyDataForPatient("allStudiesForPatient-onlyAvailableStudy-response");
        const patientId = "PID_001";
        const wrapper = shallow(
            <RegisterSubject
                patientStudyApi={patientStudyApi}
                patientRegistrationApi={patientRegistrationApi}
                patientSource={DEFAULT_PATIENT_ID_SRC}
                patientId={patientId}
                onClose={onClose}
            />
        );
        expect(spyOnMount.callCount).toEqual(1);
        expect(spyOnUpdate.callCount).toEqual(0);
        expect(spyOnFetchStudies.callCount).toEqual(1);

        expect(wrapper.state().patientId).toEqual(patientId);
        expect(wrapper.state().inError).toBe(false);
        expect(wrapper.state().isLoading).toBe(true);

        await Promise.resolve();

        expect(wrapper.state().availableStudies.length).toEqual(1);
        expect(wrapper.state().registeredStudies.length).toEqual(0);

        letWebserviceReturnStudyDataForPatient("allStudiesForPatient-onlyRegisteredStudy-response");

        wrapper.setProps({patientId: 'PID_002'});
        wrapper.update();
        expect(spyOnMount.callCount).toEqual(1);
        expect(spyOnUpdate.callCount).toEqual(2);
        expect(spyOnFetchStudies.callCount).toEqual(2);
        expect(wrapper.state().patientId).toEqual('PID_002');
        expect(wrapper.state().inError).toBe(false);
        expect(wrapper.state().isLoading).toBe(true);

        await Promise.resolve();

        expect(wrapper.state().availableStudies.length).toEqual(0);
        expect(wrapper.state().registeredStudies.length).toEqual(1);
    });

    test("Register a Patient as a Subject to a Study", async () => {
        letWebserviceReturnStudyDataForPatient("allStudiesForPatient-response");
        const PATIENT_ID = "PID_01";
        const STUDY_ID = "SID_01";
        const SUBJECT_ID = "SUB_01_REF";
        const PATIENT_ID_SRC = "HIS_01";
        const panel = mount(
            <RegisterSubject
                patientStudyApi={patientStudyApi}
                patientRegistrationApi={patientRegistrationApi}
                patientSource={PATIENT_ID_SRC}
                patientId={PATIENT_ID}
                onClose={onClose}
            />
        );
        expect(spyOnMount.callCount).toEqual(1);
        expect(spyOnUpdate.callCount).toEqual(0);
        expect(spyOnFetchStudies.callCount).toEqual(1);
        expect(spyOnRegister.callCount).toEqual(0);

        await Promise.resolve();

        expect(panel.state('patientId')).toEqual(PATIENT_ID);
        expect(panel.state('isLoading')).toBe(false);
        expect(panel.state('inError')).toBe(false);
        expect(panel.state('availableStudies').length).toEqual(1);
        expect(panel.state('registeredStudies').length).toEqual(1);

        const availableStudies = panel.find('ListStudies#availableStudies').first();
        availableStudies.prop('onSubjectRegistration')(STUDY_ID, SUBJECT_ID, DATE_OF_CONSENT);
        availableStudies.update();

        expect(spyOnRegister.callCount).toEqual(1);

        await Promise.resolve();

        expect(patientRegistrationApi.studyId).toEqual('SID_01');
        //Put in comment until E2E-630 is implemented: Only partially check the request until first name last name and birth date is implemented
        //expect(patientRegistrationApi.registerSubjectRequest).toEqual(readDataFromFile("studysubjectcontroller/registerSubject-request" + ".json"));
        tempAssertRegisterSubjectRequestUntilE2e630Implemented(patientRegistrationApi);

        expect(panel.state('patientId')).toEqual(PATIENT_ID);
        expect(panel.state('isLoading')).toBe(true);
        expect(panel.state('inError')).toBe(false);
        expect(panel.state('availableStudies').length).toEqual(0);
        expect(panel.state('registeredStudies').length).toEqual(0);
    });

    function letWebserviceReturnStudyDataForPatient(filePath) {
        const response = {data: readDataFromFile("/patientcontroller/" + filePath + ".json")};
        patientStudyApi.getAll = () => Promise.resolve(response);
    }

    function tempAssertRegisterSubjectRequestUntilE2e630Implemented(patientRegistrationApi) {
        let registerSubjectRequest = readDataFromFile("studysubjectcontroller/registerSubject-request" + ".json");
        expect(patientRegistrationApi.registerSubjectRequest.patientId).toEqual(registerSubjectRequest.patientId);
        expect(patientRegistrationApi.registerSubjectRequest.studyId).toEqual(registerSubjectRequest.studyId);
        expect(patientRegistrationApi.registerSubjectRequest.edcSubjectReference).toEqual(registerSubjectRequest.edcSubjectReference);
        expect(patientRegistrationApi.registerSubjectRequest.dateOfConsent).toEqual(registerSubjectRequest.dateOfConsent);
    }
});