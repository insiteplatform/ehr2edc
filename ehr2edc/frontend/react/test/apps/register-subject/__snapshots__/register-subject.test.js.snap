// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the component The page is rendered 1`] = `
Array [
  <div
    className="grid-block vertical shrink margin-medium-top padding-none"
  >
    <div
      className="grid-block shrink border-bottom"
    >
      <h6
        className="padding-rl padding-small-bottom margin-small-tb"
      >
        Registered EHR2EDC studies
      </h6>
    </div>
    <div
      className="grid-block vertical shrink margin-rl"
    >
      <div
        className="padding"
      >
        There are no registered studies for the patient
      </div>
    </div>
  </div>,
  <div
    className="grid-block vertical shrink margin-medium-top padding-none"
  >
    <div
      className="grid-block shrink border-bottom"
    >
      <h6
        className="padding-rl padding-small-bottom margin-small-tb"
      >
        Assign patient to EHR2EDC study 
        <i
          className="fa fa-info-circle"
          title="Click the study that you want to assign the selected patient to"
        />
      </h6>
    </div>
    <div
      className="grid-block vertical margin-bottom padding-rl"
    >
      <div
        className="padding"
      >
        There are no available studies for the patient
      </div>
    </div>
  </div>,
]
`;

exports[`Rendering the component The page is rendered with a patientId 1`] = `
Array [
  <div
    className="grid-block vertical shrink margin-medium-top padding-none"
  >
    <div
      className="grid-block shrink border-bottom"
    >
      <h6
        className="padding-rl padding-small-bottom margin-small-tb"
      >
        Registered EHR2EDC studies
      </h6>
    </div>
    <div
      className="grid-block vertical shrink margin-rl"
    >
      <div
        className="ReactTable shrink"
        style={Object {}}
      >
        <div
          className="rt-table"
          role="grid"
        >
          <div
            className="rt-thead -header"
            style={
              Object {
                "minWidth": "400px",
              }
            }
          >
            <div
              className="rt-tr"
              role="row"
            >
              <div
                className="rt-th rt-resizable-header -cursor-pointer"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "100 0 auto",
                    "maxWidth": null,
                    "width": "100px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className="rt-resizable-header-content"
                >
                  Name
                </div>
                <div
                  className="rt-resizer"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                />
              </div>
              <div
                className="rt-th rt-resizable-header -cursor-pointer"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "100 0 auto",
                    "maxWidth": null,
                    "width": "100px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className="rt-resizable-header-content"
                >
                  Description
                </div>
                <div
                  className="rt-resizer"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                />
              </div>
              <div
                className="rt-th rt-resizable-header -cursor-pointer"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "100 0 auto",
                    "maxWidth": null,
                    "width": "100px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className="rt-resizable-header-content"
                >
                  Subject ID
                </div>
                <div
                  className="rt-resizer"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                />
              </div>
              <div
                className="rt-th rt-resizable-header -cursor-pointer"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "100 0 auto",
                    "maxWidth": null,
                    "width": "100px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className="rt-resizable-header-content"
                >
                  Date Of Consent
                </div>
                <div
                  className="rt-resizer"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                />
              </div>
            </div>
          </div>
          <div
            className="rt-tbody"
            style={
              Object {
                "minWidth": "400px",
              }
            }
          >
            <div
              className="rt-tr-group"
              role="rowgroup"
            >
              <div
                className="rt-tr -odd"
                role="row"
              >
                <div
                  className="rt-td"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "100 0 auto",
                      "maxWidth": null,
                      "width": "100px",
                    }
                  }
                >
                  A Study
                </div>
                <div
                  className="rt-td"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "100 0 auto",
                      "maxWidth": null,
                      "width": "100px",
                    }
                  }
                >
                  A very serious study
                </div>
                <div
                  className="rt-td"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "100 0 auto",
                      "maxWidth": null,
                      "width": "100px",
                    }
                  }
                >
                  SUBJ_001_REF
                </div>
                <div
                  className="rt-td"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "100 0 auto",
                      "maxWidth": null,
                      "width": "100px",
                    }
                  }
                >
                  April 18, 2019
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="-loading"
        >
          <div
            className="-loading-inner"
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="grid-block vertical shrink margin-medium-top padding-none"
  >
    <div
      className="grid-block shrink border-bottom"
    >
      <h6
        className="padding-rl padding-small-bottom margin-small-tb"
      >
        Assign patient to EHR2EDC study 
        <i
          className="fa fa-info-circle"
          title="Click the study that you want to assign the selected patient to"
        />
      </h6>
    </div>
    <div
      className="grid-block vertical margin-bottom padding-rl"
    >
      <div
        className="ReactTable shrink"
        style={Object {}}
      >
        <div
          className="rt-table"
          role="grid"
        >
          <div
            className="rt-thead -header"
            style={
              Object {
                "minWidth": "235px",
              }
            }
          >
            <div
              className="rt-tr"
              role="row"
            >
              <div
                className="rt-th"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className=""
                />
              </div>
              <div
                className="rt-th rt-resizable-header -cursor-pointer"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "100 0 auto",
                    "maxWidth": null,
                    "width": "100px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className="rt-resizable-header-content"
                >
                  Name
                </div>
                <div
                  className="rt-resizer"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                />
              </div>
              <div
                className="rt-th rt-resizable-header -cursor-pointer"
                onClick={[Function]}
                role="columnheader"
                style={
                  Object {
                    "flex": "100 0 auto",
                    "maxWidth": null,
                    "width": "100px",
                  }
                }
                tabIndex="-1"
              >
                <div
                  className="rt-resizable-header-content"
                >
                  Description
                </div>
                <div
                  className="rt-resizer"
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                />
              </div>
            </div>
          </div>
          <div
            className="rt-tbody"
            style={
              Object {
                "minWidth": "235px",
              }
            }
          >
            <div
              className="rt-tr-group"
              role="rowgroup"
            >
              <div
                className="rt-tr -odd"
                onClick={[Function]}
                role="row"
              >
                <div
                  className="rt-td rt-expandable"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "35 0 auto",
                      "maxWidth": "35px",
                      "width": "35px",
                    }
                  }
                >
                  <div
                    className="rt-expander"
                  >
                    •
                  </div>
                </div>
                <div
                  className="rt-td"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "100 0 auto",
                      "maxWidth": null,
                      "width": "100px",
                    }
                  }
                >
                  Another Study
                </div>
                <div
                  className="rt-td"
                  onClick={[Function]}
                  role="gridcell"
                  style={
                    Object {
                      "flex": "100 0 auto",
                      "maxWidth": null,
                      "width": "100px",
                    }
                  }
                >
                  Another very serious study
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="-loading"
        >
          <div
            className="-loading-inner"
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
